src/python_bindings/nar/bind_nar.cpp

```diff
@@ -6,10 +6,70 @@
 #include "algorithms/nar/value_range.h"
 #include "py_util/bind_primitive.h"
 
-namespace python_bindings {
-
 namespace py = pybind11;
 
+namespace nar_serialization {
+
+py::object SerializeValueRange(std::shared_ptr<model::ValueRange> const& vr) {
+    int type_code = static_cast<int>(vr->GetTypeId());
+    if (static_cast<int>(vr->GetTypeId()) == static_cast<int>(model::TypeId::kString)) {
+        auto svr = std::dynamic_pointer_cast<model::StringValueRange>(vr);
+        return py::make_tuple(type_code, svr->domain);
+    } else if (static_cast<int>(vr->GetTypeId()) == static_cast<int>(model::TypeId::kDouble)) {
+        auto nvr = std::dynamic_pointer_cast<model::NumericValueRange<double>>(vr);
+        return py::make_tuple(type_code, py::make_tuple(nvr->lower_bound, nvr->upper_bound));
+    } else if (static_cast<int>(vr->GetTypeId()) == static_cast<int>(model::TypeId::kInt)) {
+        auto nvr = std::dynamic_pointer_cast<model::NumericValueRange<model::Int>>(vr);
+        return py::make_tuple(type_code, py::make_tuple(nvr->lower_bound, nvr->upper_bound));
+    }
+    throw std::runtime_error("Unsupported ValueRange type in serialization.");
+}
+
+std::shared_ptr<model::ValueRange> DeserializeValueRange(py::object const& obj) {
+    auto tup = obj.cast<py::tuple>();
```

Comment:
A check for tuple size makes sense here as well, pickled objects can be modified.